# **Tugas Praktikum**

### **1. Membuat model klasifikasi dengan menggunakan SVM untuk data suara, voice.csv**

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [2]:
# membaca data dari file CSV
data = pd.read_csv('data/voice.csv')
data.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [3]:
# memisahkan fitur (X) dan label (y)
X = data.drop('label', axis=1)
y = data['label']

# memisahkan data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [4]:
# membuat objek SVM
svm_classifier = SVC(kernel='linear', C=1, gamma='scale')

# melatih model SVM pada data pelatihan
svm_classifier.fit(X_train, y_train)

SVC(C=1, kernel='linear')

In [5]:
# memprediksi pada data pengujian
y_pred = svm_classifier.predict(X_test)

# menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi: {accuracy:.2f}')

# menampilkan laporan klasifikasi
classification_rep = classification_report(y_test, y_pred)
print('Laporan Klasifikasi:')
print(classification_rep)

# menampilkan matriks kebingungan
confusion_mat = confusion_matrix(y_test, y_pred)
print('Matriks Kebingungan:')
print(confusion_mat)

Akurasi: 0.94
Laporan Klasifikasi:
              precision    recall  f1-score   support

      female       0.98      0.89      0.93       297
        male       0.91      0.99      0.95       337

    accuracy                           0.94       634
   macro avg       0.95      0.94      0.94       634
weighted avg       0.94      0.94      0.94       634

Matriks Kebingungan:
[[265  32]
 [  5 332]]


### **2. Membuat model klasfikasi Multinomial Naive Bayes dengan ketentuan sebagai berikut :**
- Menggunakan data spam.csv
- Mengaktifkan stop_words menggunakan Fitur CountVectorizer
- Mengevaluasi hasilnya

In [12]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# melakukan pembacaan data dari file CSV
data = pd.read_csv('data/spam.csv', encoding='latin-1')

data.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [13]:
# menghapus kolom yang tidak digunakan
data = data.drop(columns=["Unnamed: 2", "Unnamed: 3", "Unnamed: 4"])

# mengganti label pada data
data = data.rename(columns={"v1": "labels", "v2": "sms"})

data.head()

labels                                                sms
0    ham  Go until jurong point, crazy.. Available only ...
1    ham                      Ok lar... Joking wif u oni...
2   spam  Free entry in 2 a wkly comp to win FA Cup fina...
3    ham  U dun say so early hor... U c already then say...
4    ham  Nah I don't think he goes to usf, he lives aro...

In [14]:
# melakukan pemisahan kolom sebagai label
X = data['sms']
y = data['labels']

# melakukan pemisahan data untuk dilakukan training dan testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [15]:
# membuat objek CountVectorizer dengan mengaktifkan stop words
vectorizer = CountVectorizer(stop_words='english')

# melakukan transformasi teks pesan menjadi vektor fitur
X_train = vectorizer.fit_transform(X_train)
X_test = vectorizer.transform(X_test)

In [16]:
# membuat objek Multinomial Naive Bayes
naive_bayes_classifier = MultinomialNB()

# melatih model Multinomial Naive Bayes pada data pelatihan
naive_bayes_classifier.fit(X_train, y_train)

MultinomialNB()

In [18]:
# melakukan prediksi pada data pengujian
y_pred = naive_bayes_classifier.predict(X_test)

# mlakukan perhitungan akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi: {accuracy:.2f}')

# menampilkan laporan klasifikasi
classification_rep = classification_report(y_test, y_pred)
print('Laporan Klasifikasi:')
print(classification_rep)

# menampilkan matriks kebingungan
confusion_mat = confusion_matrix(y_test, y_pred)
print('Matriks Kebingungan:')
print(confusion_mat)

Akurasi: 0.98
Laporan Klasifikasi:
              precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115

Matriks Kebingungan:
[[959   6]
 [ 12 138]]


**Evaluasi**
- Didapatkan nilai akurasi dari pelatihan serta pengujian data adalah 0.98 (98%)
- Pada laporan klasifikasi didapatkan : 
    - Nilai precission ham 0.99 dan spam 0.96
    - Nilai recall ham 0.99 dan spam 0.92
    - Nilai F1-Score ham 0.99 dan spam 0.94
- Sedangkan matriks kebingungan mencatat 12 false negative yang mana spam salah diklasifikasikan sebagai ham serta 6 false positive yang mana ham salah diklasifikasikan sebagai spam.

Dapat disimpulkan bahwa model ini bekerja dengan baik untuk membedakan antara pesan non-spam dan spam dengan tingkat akurasi sebesar 98 pesen. Walaupun masih ada beberapa klasifikasi yang salah, dengan perbandingan antara klasifikasi yang benar dan salah sebenarnya sudah cukup jauh dan akurat.

### **3. Membuat model klasfikasi Multinomial Naive Bayes dengan ketentuan,**
- Menggunakan data spam.csv
- Mengaktifkan stop_words menggunakan Fitur TF-IDF
- Mengevaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.
- Memberikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [27]:
# memisahkan antara kolom sms dan labels
X = data['sms']
y = data['labels']

# melakukan pemisahan data untuk dilakukan training dan testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [26]:
from sklearn.feature_extraction.text import TfidfVectorizer

# membuat objek TfidfVectorizer dengan mengaktifkan stop words
vectorizer = TfidfVectorizer(stop_words='english')

# melakukan transformasi teks pesan menjadi vektor fitur TF-IDF
X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)

In [25]:
# membuat objek Multinomial Naive Bayes
naive_bayes_classifier = MultinomialNB()

# melatih model Multinomial Naive Bayes pada data pelatihan dengan fitur TF-IDF
naive_bayes_classifier.fit(X_train_tfidf, y_train)

MultinomialNB()

In [24]:
# melakukan prediksi pada data pengujian
y_pred = naive_bayes_classifier.predict(X_test_tfidf)

# menghitung akurasi dengan fitur
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi dengan TF-IDF: {accuracy:.2f}')

# menampilkan laporan klasifikasi
classification_rep = classification_report(y_test, y_pred)
print('Laporan Klasifikasi dengan TF-IDF:')
print(classification_rep)

# menampilkan matriks kebingungan
confusion_mat = confusion_matrix(y_test, y_pred)
print('Matriks Kebingungan dengan TF-IDF:')
print(confusion_mat)

Akurasi dengan TF-IDF: 0.97
Laporan Klasifikasi dengan TF-IDF:
              precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115

Matriks Kebingungan dengan TF-IDF:
[[965   0]
 [ 37 113]]


**Evaluasi**
- Dari segi akurasi didapatkan nilai sebesar 0.97 atau 97% keakuratan hasil data
- Pada laporan klasifikasi didapatkan :
    - Nilai precission untuk ham sebesar 0.96 dan spam 1.00
    - Nilai recall untuk ham sebesar 1.00 dan spam 0.75
    - Nilai F1-Score untuk ham sebesar 0.98 dan spam 0.86
- Pada matriks kebingungan terdapat sejumlah 37 data yang terdeteksi false negative dan 0 data yang terdeteksi false positive.

**Kesimpulan** </br>
Dilihat dari segi keakuratan maka penggunaan Fitur CountVectorizer lebih baik daripada Fitur TF-IDF hal ini ditunjukkan dengan nilai akurasi CountVectorizer sebesar 0.98 dan TF-IDF sebesar 0.97. Serta untuk F1-Score dan recall (spam) pada laporan klasifikasi, Fitur CountVextorizer juga memiliki nilai yang lebih unggul dari pada Fitur TF-IDF. Sehingga dapat disimpulkan bahwa Fitur CountVectorizer adalah fitur terbaik untuk melakukan tugas mengklasifikasikan spam.